#1495: Objects sometimes translate when flipping

ericwa opened the pull request:
Don't snap selection bounding box center points to grid in MapViewBase::rotateObjects() or MapViewBase::doFlipObjects(), fixing https://github.com/kduske/TrenchBroom/issues/1495

Instead, round the bbox corners to integers before taking the bbox midpoint. This will give a bbox center vector whose components are either integers or offset from integers by 0.5

kduske commented:
I'm not so sure if I agree with this solution, because I don't know whether it achieves any of the goals that we have. As I see it, there are two conflicting goals:

- Avoid integer points becoming non-integer points.
- Keep the objects bounds in place.

Can you explain how your solution behaves in respect to these goals? I'm sorry, I have the flu and can't seem to think clearly right now.

ericwa commented:
No problem.. I'll try to explain my reasoning

> Keep the objects bounds in place

This will fail if the selection bounding box has non-integer min/max points, because I'm rounding the points before taking the center. 
In the more common case when the bounding box is already integer aligned, this goal is met.

> Avoid integer points becoming non-integer points.

I think this will always work. The bbox center point calculation is, for each axis, `(min + max) / 2.0`. `min` and `max` are integers, since I rounded them, so `(min + max) / 2.0` will either be an integer, or offset by 0.5 from an integer.

The key is that a center point that is 0.5 units off integer, like,  (100.5, 100.5, 100.5) is just as "safe" as an integer one, in that it will have a tidy floating-point representation where the mantissa is a prefix followed by a string of zeros.

e.g. 1.0 as a 64-bit floating point value is `1.000...(52 zeros) * 2^0`, 1.5 is stored as `1.100..(51 zeros) * 2^0`, which are both exact. Adding 0.5 to a floating point number that is an integer just adds an extra `1` bit after the last `1` in the mantissa.

Anyway, assuming the above is correct, if the flip calculation doesn't add floating-point error with an integer center point, it shouldn't add any floating point error with a 0.5-off-integer center point either.

ericwa commented:
I fixed a bug that was giving me confusing results in `MapDocumentTest.rotate` (after a 90 degree rotation, brush->bounds() was slightly off integer, but all vertices in the brush geometry were integer).

Now that's fixed, I removed the bbox rounding hack that I added earlier in this PR.

so, my comments on the goals (which the tests I added to MapDocumentTest support):
- Avoid integer points becoming non-integer points.
    - This is handled by Polyhedron::correctVertexPositions, and my argument above that it's safe to use a flip/rotation center point of say (0.5 0.5 0.5)
- Keep the objects bounds in place.
    - This will happen in all cases, because the bounds are no longer "tweaked" before getting the flip/rotation center point

ericwa commented:
In addition to removing the grid snapping of the flip center point, I removed grid snapping of the rotation center for 90 degree rotations as well... but now I'm thinking we should keep that.

Currently with this PR, rotating this by 90 degrees (grid 8 in this screenshot)
![screen shot 2016-11-13 at 4 12 06 pm](https://cloud.githubusercontent.com/assets/239161/20249733/e3f548f8-a9bc-11e6-981e-6bc98e30ba11.png)

gives this:
![screen shot 2016-11-13 at 4 12 16 pm](https://cloud.githubusercontent.com/assets/239161/20249734/e6ea0bb6-a9bc-11e6-99a0-e200e2bc2062.png)

i.e. now the block is aligned to grid 4 instead of 8.
After mapping with this for a bit, I think it's better to have things shift around after repeated rotations, but stay on grid.
